### PR TITLE
Store scancode scans locally

### PR DESCRIPTION
**Pull request by TG1999**
This PR allows to store scans locally using a PURL hash-based distribution of scans in multiple directories to avoid stressing file systems with too many files in one directory.
The goal is also to store these in Git repositories

### CODE REVIEW

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! here is some review.

File: clearcode/store_scans.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+We are using a short hash that is three characters long using hexadecimal encoding. 
+Therefore we can have 16*16*16 = 4096 repositories where each repo would contain about 
+25k scan files, if we were to store 100 million scans (which is a high mark).
+For reference one scan should use less than a 100k on average when compressed 
+with gzip or git based on looking at 15 million scans. Each repo should be roughly 
+couple hundred mega bytes big, based on 15 million scans.
+"""
+
+repo_names = [hex(hash)[2:].zfill(3) for hash in range(4096)]
```
Comment (JonoYang) [inline clearcode/store_scans.py:66]: Can you add a comment about what is going on in this list comprehension? I see that you are creating hex values of integers and then ignoring the 0x prefix, but it is not obvious at first glance.

File: clearcode/store_scans.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+def get_or_init_repo(repo_name: str, work_dir: Path, repo_namespace: str= "", pull=False):
+    """
+    Return a repo object for repo name and namespace 
+    and store it in the work dir. Clone if it does not 
+    exist optionally take the latest pull if it does exist.
+    """
+    # TODO: Manage org repo name
+    # MAYBE: CREATE ALL THE REPOS AT A TIME AND CLONE THEM LOCALLY
+    if repo_name not in get_github_repos(user_name="TG1999"):
```
Comment (JonoYang) [inline clearcode/store_scans.py:172]: Update the functions so we can pass in a username as a variable

File: clearcode/store_scans.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    if purl.namespace:
+        purl_path = purl_path / purl.namespace
+    return purl_path / purl.name / purl.version
+
+
+def get_purl_hash(purl: PackageURL, length: int=3) -> str:
+    """
+    Return a short lower cased hash of a purl.
+    """
+    #TODO: document this function in great detail, this is a part of spec
```
Comment (JonoYang) [inline clearcode/store_scans.py:158]: Do this TODO and explain how the purl hash is created

File: clearcode/store_scans.py [hunk 4]
```diff
@@ -152,38 +161,76 @@ def get_repo_name(purl: PackageURL, length: int=3) -> str:
...
+    if not response.status_code == 200:
+        raise Exception(f"HTTP {response.status_code}: Failed to get repos for {user_name}")
+
+    data = response.json()
+    for repo_data in data:
+        full_repo_name = repo_data.get("full_name")
+        if full_repo_name:
+            yield full_repo_name
 
 if __name__ == "__main__":
```
Comment (JonoYang) [inline clearcode/store_scans.py:235]: Create a management command out of this, where we can specify the directory, github org, and count as options

File: clearcode/store_scans.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    """
+    Return a short lower cased hash of a purl.
+    """
+    #TODO: document this function in great detail, this is a part of spec
+    purl_bytes = str(purl).encode("utf-8")
+    short_hash = sha512(purl_bytes).hexdigest()[:length]
+    return short_hash.lower()
+
+
+# def run(directory, repo_url):
```
Comment (JonoYang) [inline clearcode/store_scans.py]: Remove this if not needed

File: clearcode/store_scans.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+
+def get_cd_item_by_repo_name(cd_items):
+    cd_item_by_repo_name = {}
+    for cd_item in cd_items:
+        data = cd_item.data
+        cd_url = data.get("_metadata", {}).get("url")
+        coordinate = Coordinate.from_dict(coords=str2coord(cd_url))
+        if not is_valid_coordinate(coordinate):
+            #TODO: this is a bug, we need to log it
```
Comment (JonoYang) [inline clearcode/store_scans.py:110]: Print an error message here when we encounter this condition, for now.

**Review by JonoYang — CHANGES_REQUESTED** [review]
@TG1999 Sorry for the delay, I've left some comments

**Comment by pombredanne** [comment]
@TG1999 gentle ping.

**Review by pombredanne — APPROVED** [review]
Looking good!
Thanks. Merging

**Pull request merged**